### ISSUE TITLE
Timeout on event subscription

### ISSUE THREAD

**Issue by svrooij** [svrooij/node-sonos-ts#171]
I read the full discussion regarding the `timeout error handling` and I'm still not sure how to handle this case.

**Info:**

On the latest version (2.5.0), I'm getting the following error:
```
    FetchError: network timeout at: http://192.168.68.67:1400/MediaRenderer/AVTransport/Event
        at Timeout.<anonymous> (c:\dev\sonos-api\node_modules\node-fetch\lib\index.js:1484:13)
        at listOnTimeout (node:internal/timers:564:17)
        at process.processTimers (node:internal/timers:507:7) {
      type: 'request-timeout'
```

This is the only log output I got at this moment., seems like this is related to the library internal event-subscription.

Here is my `subscribe` and `unsubscribe` functions:

```js 
const subscribe = d => {
    debug("Subscribing device %s events", d.uuid)
    d.Events.on(SonosEvents.SubscriptionError, error => onSubscriptionError(d.uuid, error))
    d.RenderingControlService.Events.on(ServiceEvents.ServiceEvent, async data => await Util.async(() => onDeviceRendererChanged(d.uuid, data)))
    d.Events.on(SonosEvents.CurrentTransportStateSimple, async data => await Util.async(() => onTransportStateChanged(d, data)))
    d.ZoneGroupTopologyService.Events.on(ServiceEvents.ServiceEvent, async data => await Util.async(() => onGroupTopologyChanged(d, data)))
    d.GroupRenderingControlService.Events.on(ServiceEvents.ServiceEvent, async data => await Util.async(() => onGroupRendererChanged(d, data)))
}

const unsubscribe = d => {
    d.CancelEvents()
    d.AlarmClockService.Events.removeAllListeners(ServiceEvents.ServiceEvent)
    d.AVTransportService.Events.removeAllListeners(ServiceEvents.ServiceEvent)
    d.RenderingControlService.Events.removeAllListeners(ServiceEvents.ServiceEvent)
    d.ZoneGroupTopologyService.Events.removeAllListeners(ServiceEvents.ServiceEvent)
    d.GroupRenderingControlService.Events.removeAllListeners(ServiceEvents.ServiceEvent)
    debug('Device [%s] [%s] is now unsubsribed', d.uuid, d.name)
}
```

_Originally posted by @orensharon in https://github.com/svrooij/node-sonos-ts/issues/147#issuecomment-1277144016_

**Comment by svrooij** [comment]
The issue you commented on was closed, and while it seems to have to do with the events, it was unrelated to the notifications, so I moved this to a new issue.

@orensharon can you elaborate on this a little? Like wireless speakers? Battery powered? S1 or S2? Node version? OS? Firewall?

The `d.Events.on(...)` is the only thing where you cannot catch the error. That is because adding something to the event emitter is synchronous, but is has to subscribe for those events (which is asynchronous). As I'm writing this I could probably get around this by making a `.SubscribeForEvents(ServiceName)` method, that would solve al these issues.

**Comment by orensharon** [comment]
Hi, Sure!

First, some details about my system:
* All powered 
* Single speaker connected to the router with RJ45. All the others connected wireless.
* All updated to S2.
* Node version: 18.7.0
* OS: Windows 10. without firewall / av

I don't know if this is related, but as you can see, for each event listener I wrapped the handler with `async data => await Util.async(() => handle(d.uuid, data)))` because if I did not, any exception that `handle()` may throw is completely ignored.  

And the implementation of `Util.async(...)` it's just:

    exports.async = async function (task) {
        await task()
    }

Please let me know if there is anything else I can help you with.

**Comment by svrooij** [comment]
According to the [current code](https://github.com/svrooij/node-sonos-ts/blob/322a36d838c4cbdf424a8d2f48142d2d227ba101/src/services/base-service.ts#L397-L430), the timeout is 15 seconds at the moment. That should be plenty to subscribe for events....

I have no clue what you're doing with the `Util.async`, that seems irrelevant for now.

Can you check what happens when you try this:

```js
const subscribe = d => {
    debug("Subscribing device %s events", d.uuid)
    d.RenderingControlService.Events.on(ServiceEvents.SubscriptionError, (error) => console.log('Error subscribing', error));
    d.RenderingControlService.Events.on(ServiceEvents.ServiceEvent, (data) => console.log('New data', data));
}
```

Apart from that it might result in multiple subscriptions if you call your code like this. The following gives you the same result but less change on double event subscriptions.

```js
const subscribe = d => {
    debug("Subscribing device %s events", d.uuid)
    d.Events.on(SonosEvents.SubscriptionError, error => onSubscriptionError(d.uuid, error))
    d.Events.on(SonosEvents.RenderingControl, async data => await Util.async(() => onDeviceRendererChanged(d.uuid, data)))
    d.Events.on(SonosEvents.CurrentTransportStateSimple, async data => await Util.async(() => onTransportStateChanged(d, data)))
    // don't subscribe on ZoneTopology, let the SonosManager handle that for you. You'll only need one subscription to this service anyway, since they will all emit the same changes.
    //d.ZoneGroupTopologyService.Events.on(ServiceEvents.ServiceEvent, async data => await Util.async(() => onGroupTopologyChanged(d, data)))
    d.GroupRenderingControlService.Events.on(ServiceEvents.ServiceEvent, async data => await Util.async(() => onGroupRendererChanged(d, data)))
}
```

**Comment by orensharon** [comment]
Thanks for the replay:

I must say that the error is sporadic, I have encountered it only once.

I checked the following:

```js
const subscribe = d => {
    debug("Subscribing device %s events", d.uuid)
    d.RenderingControlService.Events.on(ServiceEvents.SubscriptionError, (error) => console.log('Error subscribing', error));
    d.RenderingControlService.Events.on(ServiceEvents.ServiceEvent, (data) => console.log('New data', data));
}
```

And the output is:

    sonos:device-discovery Device discovery created +0ms
    sonos:sonosmanager InitializeWithDiscovery timeout: 10 +0ms
    sonos:device-discovery Search one device in 10 seconds +3ms
    sonos:device-discovery sendBroadcast() +0ms
    sonos:device-discovery UDP port not bound, waiting 100ms +1ms
    sonos:device-discovery UDP socket started listening +0ms
    sonos:device-discovery Bound to port 60938 +1ms
    sonos:device-discovery sendBroadcast() +108ms
    sonos:device-discovery SSDP message
    sonos:device-discovery HTTP/1.1 200 OK
    CACHE-CONTROL: max-age = 1800
    EXT:
    LOCATION: http://192.168.68.57:1400/xml/device_description.xml
    SERVER: Linux UPnP/1.0 Sonos/69.1-33120 (ZPS22)
    ST: urn:schemas-upnp-org:device:ZonePlayer:1
    USN: uuid:RINCON_5CAAFDCFAA1801400::urn:schemas-upnp-org:device:ZonePlayer:1
    X-RINCON-HOUSEHOLD: Sonos_XXXXXXXXXXXXXXXXXXXXXX
    X-RINCON-BOOTSEQ: 33
    BOOTID.UPNP.ORG: 33
    X-RINCON-PROXY: RINCON_542A1B8D356301400
    X-RINCON-WIFIMODE: 0
    X-RINCON-VARIANT: 2
    HOUSEHOLD.SMARTSPEAKER.AUDIO: Sonos_XXXXYXXXXXXXXXXXXXXXXXXXXX.XXXXXXXXYEXXXXXXXX-X
    SECURELOCATION.UPNP.ORG: https://192.168.68.57:1443/xml/device_description.xml
    X-SONOS-HHSECURELOCATION: https://192.168.68.57:1843/xml/device_description.xml +33ms
    sonos:device-discovery Found ZPS22 on 192.168.68.57 +4ms
    sonos:sonosmanager Discovery found player with ip: 192.168.68.57 +155ms
    sonos:sonosmanager Initialize() +1ms
    sonos:service:zonegrouptopology:192.168.68.57 GetZoneGroupState() +0ms
    sonos:service:zonegrouptopology:192.168.68.57 Listener added for serviceEvent  (sid: 'undefined', SONOS_DISABLE_EVENTS: true) +31ms
    sonos:service:zonegrouptopology:192.168.68.57 Subscribing to events +1ms
    sonos:eventlistener Listener endpoint: http://192.168.68.52:6329/sonos/{sonos-uuid}/{serviceName} +0ms
    sonos:service:zonegrouptopology:192.168.68.57 Creating event subscription with callback: http://192.168.68.52:6329/sonos/58e66915-d14e-48f6-dd01-29c4ac64e49a/ZoneGroupTopology +7ms
    sonos:service:zonegrouptopology:192.168.68.68 GetZoneGroupState() +0ms
    sonos:eventlistener Starting event listener on port 6329 +17ms
    sonos:eventlistener Got event on /sonos/58e66915-d14e-48f6-dd01-29c4ac64e49a/ZoneGroupTopology SID: uuid:RINCON_5CAAFDCFAA1801400_sub0000000217 +14ms
    sonos:service:zonegrouptopology:192.168.68.57 Got event +47ms
    sonos:service:zonegrouptopology:192.168.68.54 GetZoneGroupAttributes() +0ms
    sonos:service:zonegrouptopology:192.168.68.61 GetZoneGroupAttributes() +0ms
    sonos:service:zonegrouptopology:192.168.68.68 GetZoneGroupAttributes() +46ms
    sonos:service:zonegrouptopology:192.168.68.67 GetZoneGroupAttributes() +0ms
    sonos:device:192.168.68.54 Listener added for subscriptionError (isSubscribed: false) +0ms
    sonos:service:renderingcontrol:192.168.68.54 Listener added for serviceEvent  (sid: 'undefined', SONOS_DISABLE_EVENTS: true) +0ms
    sonos:service:renderingcontrol:192.168.68.54 Subscribing to events +0ms
    sonos:service:renderingcontrol:192.168.68.54 Creating event subscription with callback: http://192.168.68.52:6329/sonos/RINCON_542A1B8D356301400/RenderingControl +1ms
    sonos:device:192.168.68.61 Listener added for subscriptionError (isSubscribed: false) +0ms
    sonos:service:renderingcontrol:192.168.68.61 Listener added for serviceEvent  (sid: 'undefined', SONOS_DISABLE_EVENTS: true) +0ms
    sonos:service:renderingcontrol:192.168.68.61 Subscribing to events +1ms
    sonos:service:renderingcontrol:192.168.68.61 Creating event subscription with callback: http://192.168.68.52:6329/sonos/RINCON_7828CA237E9601400/RenderingControl +0ms
    sonos:device:192.168.68.68 Listener added for subscriptionError (isSubscribed: false) +0ms
    sonos:service:renderingcontrol:192.168.68.68 Listener added for serviceEvent  (sid: 'undefined', SONOS_DISABLE_EVENTS: true) +0ms
    sonos:service:renderingcontrol:192.168.68.68 Subscribing to events +0ms
    sonos:service:renderingcontrol:192.168.68.68 Creating event subscription with callback: http://192.168.68.52:6329/sonos/RINCON_5CAAFDB8283E01400/RenderingControl +0ms
    sonos:device:192.168.68.67 Listener added for subscriptionError (isSubscribed: false) +0ms
    sonos:service:renderingcontrol:192.168.68.67 Listener added for serviceEvent  (sid: 'undefined', SONOS_DISABLE_EVENTS: true) +0ms
    sonos:service:renderingcontrol:192.168.68.67 Subscribing to events +0ms
    sonos:service:renderingcontrol:192.168.68.67 Creating event subscription with callback: http://192.168.68.52:6329/sonos/RINCON_7828CAB4010601400/RenderingControl +1ms
    sonos:eventlistener Got event on /sonos/RINCON_542A1B8D356301400/RenderingControl SID: uuid:RINCON_542A1B8D356301400_sub0000005475 +64ms
    sonos:service:renderingcontrol:192.168.68.54 Got event +26ms
    sonos:eventlistener Got event on /sonos/RINCON_7828CAB4010601400/RenderingControl SID: uuid:RINCON_7828CAB4010601400_sub0000005233 +10ms
    sonos:service:renderingcontrol:192.168.68.67 Got event +18ms
    sonos:eventlistener Got event on /sonos/RINCON_7828CA237E9601400/RenderingControl SID: uuid:RINCON_7828CA237E9601400_sub0000005230 +4ms
    sonos:service:renderingcontrol:192.168.68.61 Got event +36ms
    sonos:eventlistener Got event on /sonos/RINCON_5CAAFDB8283E01400/RenderingControl SID: uuid:RINCON_5CAAFDB8283E01400_sub0000005229 +8ms
    sonos:service:renderingcontrol:192.168.68.68 Got event +43ms

Regarding the multiple subscriptions of `ZoneGroupTopologyService` - 

I understand now, that only single device should be registered to this event, I will give it a try.
Is this also applies to `d.GroupRenderingControlService.Events.on(...)` subscription?

**Comment by orensharon** [comment]
Hi,

Update - 
@svrooij, I think this issue can be closed.
I missed that `manager.CheckAllEventSubscriptions()` is an async function.

<hr/>

**Background:**

For some reason, after a while, I'm no longer receiving events (such as `TransportState` and `RenderingControl`) from devices.
(I don't have any debug outputs for that yet) 

I was able to solve this by using the code I found in one of the examples
https://github.com/svrooij/node-sonos-ts/blob/322a36d838c4cbdf424a8d2f48142d2d227ba101/examples/events.js#L36

I called the `manager.CheckAllEventSubscriptions()` manually and repeatedly,
and I forgot to use the `async await` - this is the cause of the `timeout error`.

<hr/>

**Questions/Suggestion:**

I saw that the `renewEventSubscription()` inside the `base-service` is triggered every 10 minutes.
https://github.com/svrooij/node-sonos-ts/blob/322a36d838c4cbdf424a8d2f48142d2d227ba101/src/services/base-service.ts#L425

* Can this interval be configurable?
Is there a minimum limit for this value?

* Is there a way to know that `renewEventSubscription()` re-subscribed after the subscription was lost?

To simulate the `renewEventSubscription` success/fail, I turned off and on my network adapter (Wifi) and it took 10 minutes to receive the `SubscriptionError` error or for the re-subscription to take place.

**Comment by svrooij** [comment]
I would love to implement a way to "monitor" devices. At the moment to my knowledge it does not sends out events if some speaker is no longer connected. I think this can only be done by continuedly doing a call to some endpoint like this https://github.com/svrooij/node-sonos-ts/blob/322a36d838c4cbdf424a8d2f48142d2d227ba101/src/sonos-device.ts#L187.

I'm not sure what kind of stress that would be on the speakers if I would get that information like every 30 seconds, just to check if the speaker is still online.

What do you think of something like on the `SonosDevice` there is a timer that checks this endpoint, but is get reset if there is any event received on that speaker. So as long as it's playing and sending regular updates nothing happens, but as soon as there are no longer updates we check that endpoint like every 30 seconds and emit an event if a speaker can no longer be found.

**Comment by orensharon** [comment]
Thank you again for the response.

I want to separate between 2 cases please:

1. Speaker is online, but lost its subscription. re-subscription is required.

    This is the reason why I have initially tried to call `manager.CheckAllEventSubscriptions()` inside a timer.
As I mentioned, sometimes I no longer receive events from speakers (it happens when I regroup speakers frequently as part of a debugging session).

    In this case I want to re-subscribe the speaker as soon as possible.
In my solution, the interval is 5 seconds ( I'm not sure what the stress is here):

```js
const startMonitor = (rate = 5000) => {
    _monitor = setInterval(async () => {
        try {
            await _manager.CheckAllEventSubscriptions()
        } catch (error) {
            logger.warn('Check all event subscriptions failed: %s', error.message)
        }
    }, rate)
}
```
2. Speaker got offline (powered off, disconnected from network, etc...)

>What do you think of something like on the SonosDevice there is a timer that checks this endpoint, but is get reset if there is any event received on that speaker. So as long as it's playing and sending regular updates nothing happens, but as soon as there are no longer updates we check that endpoint like every 30 seconds and emit an event if a speaker can no longer be found.

If I understand you correctly, when the speakers are idle (or playing) - the timer will check the endpoint every 30 secs?

<hr />

Do you know how the Sonos app is dealing with those scenarios?

I did try to check it, I don't have unequivocal findings, but I did notice that it may take the app several minutes to remove the speaker from the list, but sometimes it's marked as 'Offline' immediately.

I'm trying to find a way to "disconnect" the speaker without powering it off every time to make more tests, any ideas?

**Comment by svrooij** [comment]
> Speaker is online, but lost its subscription. re-subscription is required.

I never noticed this, this seems to be related to you changing the speakers often (or them not having a static IP).

The only reason I can think of this would happen is, if the speaker is turned off or lost wifi connection. But that should be discovered by the timer (if it takes longer then 30 seconds)

> Timer every 30 seconds when idle

The is exactly the result of this would be implemented, (and also the reason why it isn't there yet).

You can always call the check all subscriptions method as often as you wish, but I cannot oversee the issues that would give in the long-run.

**Comment by orensharon** [comment]
> this seems to be related to you changing the speakers often (or them not having a static IP).

* I did change the speakers often (created new groups fast) to reproduce the problem.
* IP address is reserved for each speaker.


> You can always call the check all subscriptions method as often as you wish

So if I get you - the new timer will only check the connectivity to the speaker (like ping).
And if a subscription is getting lost for any reason, I have to handled it manually (such re-subscribe using the `CheckAllEventSubscriptions()`), or wait for it to handled internally by your code (the 10 mins timer)

**Comment by svrooij** [comment]
If the ping discovered a speaker is not available anymore, it might also discover the speaker is available again and then restart the event subscription.

Just airing ideas for this new feature.

**Comment by svrooij** [comment]
@orensharon is this issue still relevant? And if so, can you summarize the proposed solution?
